### ISSUE TITLE
LazyCache with AutoFac as DI

### ISSUE THREAD

**Issue by sw-ms-pradipbaria** [alastairtree/LazyCache#82]
Hi @alastairtree ,

I am getting below error and not getting on how to fix it, we are using Autofac for DI.

> _Autofac.Core.DependencyResolutionException_

> **Cannot choose between multiple constructors with equal length 1 on type 'LazyCache.CachingService'. Select the constructor explicitly, with the UsingConstructor() configuration method, when the component is registered.**

Registered in startup like below

![image](https://user-images.githubusercontent.com/31621470/61796542-35ca4100-ae43-11e9-8795-ce9e1df53d6b.png)

And injecting in Repositoy like below

![image](https://user-images.githubusercontent.com/31621470/61796450-0ddadd80-ae43-11e9-9aa7-2fda8bcea1c3.png)

**Comment by AlexBar** [comment]
The same issue here. The root cause is that Autofac can't choose between 3 constructors because it can automatically inject Lazy/Func even if it is not registered as Lazy or Func:

```
        public CachingService(Lazy<ICacheProvider> cacheProvider)
        {
            this.cacheProvider = cacheProvider ?? throw new ArgumentNullException(nameof(cacheProvider));
        }

        public CachingService(Func<ICacheProvider> cacheProviderFactory)
        {
            if (cacheProviderFactory == null) throw new ArgumentNullException(nameof(cacheProviderFactory));
            cacheProvider = new Lazy<ICacheProvider>(cacheProviderFactory);
        }

        public CachingService(ICacheProvider cache) : this(() => cache)
        {
            if (cache == null) throw new ArgumentNullException(nameof(cache));
        }
```

**Comment by TimSirmovics** [comment]
This looks to be related to #75.

**Comment by alastairtree** [comment]
I think this should be fixed in 2.0.4 as i have made the constructor selection logic explicit. Can you give it  a test and let me know? Thanks

**Comment by alastairtree** [comment]
Closing as probably fixed, file a new issue asking to reopen this if required.
